provisioning/containers/gui-common/websockify/instmetrics.go

```diff
@@ -0,0 +1,197 @@
+	cpuLimit string
+	// Application container resources.limits.memory.
+	memoryLimit string
+	// PodName where application container is running.
+	podName string
+	// Last Resources extracted from CustomMetricsServer
+	cachedResources      *Resources
+	cachedResourcesMutex sync.RWMutex
+	// <UID, ConnInfo> map
+	connectionsTracking *sync.Map
+}
+
+// ServeHTTP serves WS server.
+func (h *InstanceMetricsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Println("upgrade error:", err)
+		return
+	}
+
+	var updatePeriod int64 = 2
+	if up, ok := r.URL.Query()["updatePeriod"]; ok {
+		updatePeriod, err = strconv.ParseInt(up[0], 10, 16)
+		if err != nil {
+			http.Error(w, "Invalid updatePeriod: it must be Integer (update seconds)", http.StatusBadRequest)
+			return
+		}
+	}
+	updatePeriodD := time.Duration(updatePeriod) * time.Second
+
+	var connUID *string
+	if uid, ok := r.URL.Query()["connUid"]; ok {
+		// noVNC page request
+		uids := uid[0]
+		connUID = &uids
+	} else {
+		// metricsDashboard request
+		connUID = nil
+	}
+
+	ip := r.Header.Get(headerXForwardedFor)
+	log.Printf("Incoming websocket connection on path /usages with update period %d from IP=%s", updatePeriod, ip)
+
+	go h.sendInstMetricsCycle(ws, updatePeriodD, connUID)
+}
+
+func (h *InstanceMetricsHandler) sendInstMetricsCycle(ws *websocket.Conn, updatePeriod time.Duration, connUID *string) {
+	ticker := time.NewTicker(updatePeriod)
+	for range ticker.C {
+		ctx, cancel := context.WithTimeout(context.Background(), updatePeriod)
+
+		if !h.cachedResourcesIsUpdated(updatePeriod) {
+			if err := h.updateCachedResources(ctx, updatePeriod); err != nil {
+				log.Printf("[sendCustomMetricsCycle] Error sending data on websocket. Closing WebSocket connection. Error: %v", err)
+				cancel()
+				ticker.Stop()
+			}
+		}
+		if err := h.sendCachedResources(ws, updatePeriod, connUID); err != nil {
+			log.Printf("[sendCustomMetricsCycle] Error sending data on websocket. Closing WebSocket connection. Error: %v", err)
+			cancel()
+			ticker.Stop()
+		}
+
+		// Operation running the context completed.
+		cancel()
+	}
+}
+
+// cachedResourcesIsUpdated returns true if chachedResources is no older than updatePeriod.
+func (h *InstanceMetricsHandler) cachedResourcesIsUpdated(updatePeriod time.Duration) bool {
+	h.cachedResourcesMutex.RLock()
+	defer h.cachedResourcesMutex.RUnlock()
+	if cr := h.cachedResources; cr == nil {
+		return false
+	}
+	return time.Since(h.cachedResources.Timestamp) < updatePeriod
+}
+
+// updateCachedResources retrieves updated ContainerMetrics from instMetrics server.
+func (h *InstanceMetricsHandler) updateCachedResources(ctx context.Context, updatePeriod time.Duration) error {
+	h.cachedResourcesMutex.RLock()
+	if cr := h.cachedResources; cr != nil && time.Since(h.cachedResources.Timestamp) < updatePeriod {
+		h.cachedResourcesMutex.RUnlock()
+		return nil
+	}
+	h.cachedResourcesMutex.RUnlock()
```

Comment:
```suggestion
	cr := h.cachedResources;
	h.cachedResourcesMutex.RUnlock()
	if cr != nil && time.Since(cr.Timestamp) < updatePeriod {
		return nil
	}
```
Doesn't this work too?

Reply:
Yes, It does. Thanks for the advice